### PR TITLE
fix!: leftovers work for sync fallback

### PR DESCRIPTION
**Pull request by qfrank**
This is a continue work to [PR](https://github.com/status-im/status-go/pull/5614#issue-2434423574)
2 Major changes in this PR:
- removed the compatibility with v2.29, in other word, version below v2.30 won't be able to parse the sync code generated with this PR
- add AC notifications when initiating the sync fallback

relate mobile [PR](https://github.com/status-im/status-mobile/pull/21090)

### CODE REVIEW

File: .gitignore
Comment (igor-sirotin) [inline .gitignore]: something new, where is it coming from? 🤔

File: protocol/messenger.go
Comment (igor-sirotin) [inline protocol/messenger.go]: Typo `// // `?

Comment (igor-sirotin) [inline protocol/messenger.go]: Actually, are we sure we want to delete the notification, no like `Accept` it or smth?

File: protocol/messenger.go
Comment (igor-sirotin) [inline protocol/messenger.go]: Maybe `else if` here? Will be less spaghetti.

File: protocol/messenger_pairing_and_syncing.go
Comment (igor-sirotin) [inline protocol/messenger_pairing_and_syncing.go]: This looks like a duplication from `saveDataAndPrepareResponse`. 
Can you please extract it to a func?

**Comment by igor-sirotin** [comment]
cc @glitchminer looks like this is a good case for some user stories 🙂

File: protocol/messenger.go
Comment (jrainville) [reply protocol/messenger.go]: As we wish. I think the notification serves no further purpose once accepted, but it's as we prefer.

File: protocol/messenger.go
Comment (jrainville) [reply protocol/messenger.go]: Good point, not sure why I did it like this.

**Review by jrainville — COMMENTED** [review]
Looks good, though more than half the code is mine :see_no_evil: 

~~@qfrank do you want to address Igor's comments or do you want me to do it?~~
nvm I did it :D

**Comment by jrainville** [comment]
I added a commit with the review comment fixes. I also rebased and updated my status-desktop PR here: https://github.com/status-im/status-desktop/pull/15820

It works again as expected

File: .gitignore
Comment (qfrank) [reply .gitignore]: generated after `make lint-fix`

Comment (ilmotta) [reply .gitignore]: @qfrank, I couldn't generate these files, which makes sense to me because `.clj-kondo` is the directory generated by the Clojure linter. I don't know how it got generated in status-go. In which case we can remove these changes from `.gitignore`. The directory `.lsp/` I also couldn't generate with `make lint-fix` in status-go.

**Review by ilmotta — APPROVED** [review]

File: .gitignore
Comment (qfrank) [reply .gitignore]: image

It's weird as I can't generate it with `make lint-fix` now, but I can remember it was generated with `make lint-fix`, The evidence is in the screenshot. These temporary files are still in my local project. 😅

Comment (ilmotta) [reply .gitignore]: They have wings and flew from status-mobile to status-go 🪰

**Comment by igor-sirotin** [comment]
@qfrank @jrainville are you still around this PR?
Looks like it needs some love around commit messages 😄

**Comment by qfrank** [comment]
> @qfrank @jrainville are you still around this PR? Looks like it needs some love around commit messages 😄

updated @igor-sirotin

**Comment by jrainville** [comment]
@qfrank feel free to squash my commits if you want

**Comment by Parveshdhull** [comment]
Hi @qfrank, rebased the PR - https://github.com/status-im/status-mobile/pull/21090#issuecomment-2349082865

File: protocol/messenger.go
Comment (igor-sirotin) [inline protocol/messenger.go]: Less spaghetti please 🙏 

```suggestion
	if notification == nil {
	    return nil
	}
	
    updatedAt := m.GetCurrentTimeInMillis()
    notification.UpdatedAt = updatedAt
    notification.Deleted = true
    // we shouldn't sync deleted notification here,
    // as the same user on different devices will receive the same message(CommunityCancelRequestToJoin) ?
    err = m.persistence.DeleteActivityCenterNotificationByID(types.FromHex(installationID), updatedAt)
    if err != nil {
	    m.logger.Error("failed to delete notification from Activity Center", zap.Error(err))
	    return err
    }
    
    // sending signal to client to remove the activity center notification from UI
    response.AddActivityCenterNotification(notification)

	return nil
```

File: services/ext/api.go
Comment (igor-sirotin) [inline services/ext/api.go:1071]: As I understand, this is the only breaking change in this PR.

If you really think we should rename it, then:
1. Mark `EnableAndSyncInstallation` as deprecated, but keep it working.
2. Introduce `EnableInstallationAndSync` and update clients to use the new endpoint. 

This way we can ensure clients are still compatible with the changes.

Comment (jrainville) [reply services/ext/api.go:1071]: I'm not fully aware of all the implications, but as I understood it, the breaking change it bigger than that. Once this is released, people will only be able to pair between 2.31 and up.

I guess it was not possible to keep backwards compatibility, otherwise Andrea would have thought of it.

As for renaming the API, both desktop and mobile are creating PRs at the same time, so I think it's fine to just rip the band aid right now and not keep an unused API?

Comment (igor-sirotin) [reply services/ext/api.go:1071]: I think it depends on the terminology of a `breaking change` that we should agree upon.
Let me discuss this with the @status-im/status-go-guild today, I'll be back with an answer in the evening.

**Review by Samyoul — APPROVED** [review]
Thank you @qfrank , I've gone through this and it all seems really nicely written.

File: services/ext/api.go
Comment (Samyoul) [reply services/ext/api.go:1071]: We have 3 types of breaking change:

- Breaking between versions (the type that we want to avoid the most)
- Breaking of functionality (basically introducing a (or several) bug(s))
- Breaking between `status-go` and client (potentially this case)

I think breaks between `status-go` and clients can be managed, as long as they are managed they are only a problem if someone has an old branch and doesn't rebase against the latest `develop`.

Comment (qfrank) [reply services/ext/api.go:1071]: I have added the old API back, can I get your approval now? 🙂 Your approval is valuable for me! @igor-sirotin

Comment (igor-sirotin) [reply services/ext/api.go:1071]: Indeed, this would be considered as breaking change, because we're breaking compatibility with previous versions 👍 

Yet, I think it's better to avoid such functions renaming in general.

Comment (qfrank) [reply services/ext/api.go:1071]: Regarding "breaking compatibility with previous versions", we have had many discussions before. I just created this PR according to the original conclusion. Regarding breaking change, I never want to produce breaking change unless I have to... This PR is a historical leftover task.

Regarding function renaming, I have a different opinion. The naming of the old API really makes it difficult to understand its intention well. At that time, I only adopted the feedback from PR review. But looking back now, this naming is not so clear, so there is a new naming. If we can communicate with relevant mobile/desktop people in advance when renaming and retain the old function, it is not impossible. I will create an issue to follow up and remove the old function. @igor-sirotin cc @ilmotta @jrainville

Comment (igor-sirotin) [reply services/ext/api.go:1071]: @qfrank I'm not really saying that we should stick to the old function name. 
But there's a way to make this carefully: create a new function and deprecate the old one. Then the switch to the new function name will be seamless 🙌

Comment (qfrank) [reply services/ext/api.go:1071]: > create a new function and deprecate the old one

I've done this yesterday 🙂

Comment (qfrank) [reply services/ext/api.go:1071]: Can I merge it now? I'm waiting for your approval 🥹 @igor-sirotin

Comment (igor-sirotin) [reply services/ext/api.go:1071]: Oh! I only read your comment and and didn't check the code 😄

**Review by igor-sirotin — APPROVED** [review]

File: services/ext/api.go
Comment (qfrank) [reply services/ext/api.go:1071]: no worries, we can create another PR if needed

**Pull request merged**
